### DISCUSSION TITLE
How to create a cluster?

### DISCUSSION THREAD

**Discussion by SSaiPrasad7** [emqx/emqx#4595] [Q&A] [answered]
Hello @zmstone,
I would like to create emqx cluster with two machines i.e.
1. Local machine(Ubuntu 20.04,emqx version:4.3-rc.3-0a092c8b)
2. AWS instance(Ubuntu 18.04,emqx version: 4.29)

How to connect these two emqx servers?

This is the error message when I am trying to join the the server to cluster.

![image](https://user-images.githubusercontent.com/36000620/114981902-51456d00-9eac-11eb-8fac-cbd60dd211a1.png)

**Comment by zmstone** [comment, marked as answer]
Good chance that the nodes can not communicate each other.
Please refer to this doc:
https://docs.emqx.io/en/broker/v4.3/getting-started/cluster.html#firewall-settings

**Reply by SSaiPrasad7** [reply]
@zmstone 
I read the docs but I did not understand configuration of firewall settings under `emqx.conf` file .
I request you tell me how to connect the servers (to form a cluster) from starting till the end. So, It will be beneficial for others like me.

So far, I checked the firewall status by `sudo ufw status`. It is `inactive` on both servers.
I changed the `localhost` to public IP address.

Check this image:
![image](https://user-images.githubusercontent.com/36000620/115023940-0abc3680-9edd-11eb-9e2c-ea6ff2069dfc.png)

**Reply by k32** [reply]
Hi! The IP address of the local ubuntu machine (`emqx@192.168.224.129`) looks like it belongs to the internal network. While address of the other machine looks like external. Can you run `echo $SSH_CLIENT` command on the AWS machine?

Most likely the AWS machine is not able to talk to the local machine, because in a typical network setup it is behind NAT.

**Reply by SSaiPrasad7** [reply]
Hi @k32 
![image](https://user-images.githubusercontent.com/36000620/115027910-eca50500-9ee1-11eb-908c-f310ae3dba46.png)

**Reply by k32** [reply]
Right. As expected, your local machine is behind NAT. 
EMQX uses mesh topology for the cluster, meaning each node of the cluster should be able to connect to all the other nodes. But when one of the nodes is behind NAT, the other nodes in the cluster cannot connect to it. Therefore, all EMQX nodes should run in the same network.

Are you doing some test setup?

**Reply by SSaiPrasad7** [reply]
Thank you. I understood the concept about Emqx clustering.
@k32 test setup? I am sorry. I did not get you.

**Reply by k32** [reply]
Sorry, what I mean is: are you trying to create a temporary cluster for testing?

**Reply by SSaiPrasad7** [reply]
Yes. I am

**Reply by k32** [reply]
Ok. The easiest and the most robust option would be to run both EMQX nodes either in AWS or locally. For the second option one can use docker-compose: https://docs.emqx.io/en/broker/v4.3/getting-started/install.html#install-emq-x-in-docker-contain-a-simple-docker-compose-cluster

**Reply by SSaiPrasad7** [reply]
Okay. Thank you, I will connect two emqx nodes locally. May I know the steps(procedure) to connect the nodes?

**Reply by k32** [reply]
docker-compose script does that automatically, you just need to install docker-compose in your Ubuntu VM.

**Reply by SSaiPrasad7** [reply]
Fine. I will do now

**Reply by SSaiPrasad7** [reply]
Hello @k32, 
Done with Docker as well as Emqx setup. I am getting an error when I am running `docker-compose -p my_emqx up -d`


![image](https://user-images.githubusercontent.com/36000620/115043956-a86e3080-9ef2-11eb-984d-5ccb4d0a8ad3.png)

**Reply by k32** [reply]
`docker-compose` comes in a separate package, you can install it using `sudo apt install docker-compose` command

**Reply by SSaiPrasad7** [reply]
```
Cluster status: #{running_nodes => ['emqx@node1.emqx.io','emqx@node2.emqx.io'],
                  stopped_nodes => []}
```
I got this output but I am able to see the one node on the dashboard

![image](https://user-images.githubusercontent.com/36000620/115056707-cb9fdc80-9f00-11eb-88d1-92c251992e36.png)

**Reply by k32** [reply]
This status means there are no stopped nodes, so I guess it's expected?

**Reply by SSaiPrasad7** [reply]
How to publish and subscribe to a particular node ?

**Reply by k32** [reply]
Each EMQ X docker container has its own IP in your local network, which can be found via `docker inspect` command. You can connect the client to one of these IPs.
